Navneeth add google doc link to weekly summaries email

navneeeth opened the pull request:
# Description
![task description](https://github.com/OneCommunityGlobal/HGNRest/assets/44264834/807b989f-3e0f-41e2-ab2a-29e5b1197584)

## Related PRs (if any):
None. Does not require the Frontend PR to run.

## Main changes explained:
- Fetch the admin links from the weeklySummaries results.
- Retrieve the Google Doc Link from the admin links if available,
- Update the email body to include Google Doc Link in the Weekly Summaries.
- Remove Sangam's email from the recipients list.

## How to test:
1. Check into the current branch
2. Ensure that Nodemailer is set up (see the Notes section) so that you have the functionality to send an email from the app
3. Make the following code changes in `src/cronjobs/userProfileJobs.js`:
- Un-comment line 8 `// '* * * * *', // Comment out for testing. Run Every minute.` 
This is to run the cronjob every minute for convenient testing. Please ensure that you are stopping the execution of the code immediately after an email is sent. It is advisable to set up custom schedule values (see the Notes section) that are based on the time and date when you are testing the PR.
- Comment line 9 `'1 0 * * 0', // Every Sunday, 1 minute past midnight.`
-  Comment line 13 `await userhelper.assignBlueSquareForTimeNotMet();`
- Comment line 14 `await userhelper.applyMissedHourForCoreTeam();`
- Comment line 16 `await userhelper.deActivateUser();`
- Comment line 17 `await userhelper.deleteExpiredTokens();`
- Comment line 19 `await userhelper.awardNewBadges();`
- Comment line 20 `await userhelper.reActivateUser();`
- Comment line 21 `await userhelper.deActivateUser();`
- Set the day of the week in line 11 `const SUNDAY = 0;` as the integer day when you will be testing this PR where counting begins from 0 and the week begins from Sunday (0 for Sunday, 1 for Monday, and so on...)
4. Make the following code changes in `src/helpers/userhelper.js`:
- Comment line 273 `"onecommunityglobal@gmail.com, onecommunityhospitality@gmail.com",`
- Add a line below it with your email in the same format as above: `"youremail@email.com",`    
5. Save the changes in both the files, run `npm install`, and run `npm run dev` to run this PR locally
6. Once you receive the email, verify if the Google Doc Link is a field added to the Weekly Summaries email (see the Notes section for the expected format)
7. MOST IMPORTANT: Immediately stop the execution.

## Test Video (with instructions narrated):

https://github.com/OneCommunityGlobal/HGNRest/assets/44264834/5fa89add-57b2-426c-8628-816ea156cc90

## Notes
### Syntax for setting schedule values in a cronjob
![Cron Job syntax](https://github.com/OneCommunityGlobal/HGNRest/assets/44264834/b97b0b4d-8e40-430c-ba0a-ade5d9036f5f)
- Please follow [this guide](https://www.ibm.com/docs/en/db2oc?topic=task-unix-cron-format) to learn more about the UNIX cron format.

### Google Doc Link in the received email (expected format)
![google doc link in the received email](https://github.com/OneCommunityGlobal/HGNRest/assets/44264834/28c87fb4-ac8b-4ec8-acca-b2384eb3733e)

### Nodemailer setup
You need to have Nodemailer set up to add the following environment variables to the `.env` file in your backend setup in order to send an email:
```
REACT_APP_EMAIL
REACT_APP_EMAIL_CLIENT_ID
REACT_APP_EMAIL_CLIENT_SECRET
REACT_APP_EMAIL_CLIENT_REDIRECT_URI
REACT_APP_EMAIL_REFRESH_TOKEN
```

- Set your sender email ID as the value for the REACT_APP_EMAIL environment variable.
- Please follow this tutorial to set up Nodemailer and OAuth2 to set up and add the Client ID, Client Secret, Redirect URI, and Refresh Token: [Link to YouTube Video](https://youtu.be/-rcRf7yswfM?si=I9LdYTpn60EVVc3a)
- Ensure that the following environment variable with value is added to the `.env` file:
```sendEmail=TRUE```

abdel-lall commented on src/helpers/userHelper.js:
```diff
@@ -171,6 +172,14 @@ const userHelper = function () {
           ? `<a href="${mediaUrl}">${mediaUrl}</a>`
           : 'Not provided!';
 
+        const googleDocLinkValue = adminLinks 
```
This line: const googleDocLinkValue = adminLinks ? ... : .....
SHoul be: const googleDocLinkValue = adminLinks.length > 0 ? ... : .... 
Because the  adminLinks variable is an array. Even when it is empty, it still evaluates to truthy. This means that the ternary operation will always return true.

navneeeth replied:
Hello, @abdel-lall! Thank you for your review! I agree with your suggestions and have maded the necessary changes to the implementation, Requesting your approval upon testing the updated PR. Thank you!

abdel-lall commented on src/helpers/userHelper.js:
```diff
@@ -171,6 +172,14 @@ const userHelper = function () {
           : 'Not provided!';
 
+        const googleDocLinkValue = adminLinks 
+        ? adminLinks.find(link => link.Name === 'Google Doc') 
```
I have a suggestion for an edit:
**This line:** adminLinks.find(link => link.Name === 'Google Doc')
**Should be:** adminLinks.find(link => link.Name === 'Google Doc' &&  link.Link)
A Check for Link value is needed in case "Google Doc" name exist but the Link is empty.
Ex:
```
"adminLinks": [
    {
      "Name": "Google Doc",
      "Link": ""
    },]
```
This adjustment resolves the issue shown in the picture.
 
![Screenshot 2024-02-07 184250](https://github.com/OneCommunityGlobal/HGNRest/assets/49083865/775daf5d-b21e-4528-9209-a8cc9c82803b)

navneeeth replied:
Hello, @abdel-lall! Thank you for your review! I agree with your suggestions and have maded the necessary changes to the implementation, Requesting your approval upon testing the updated PR. Thank you!

abdel-lall reviewed (approved):
Works great, nice job!
![Screenshot 2024-02-08 222209](https://github.com/OneCommunityGlobal/HGNRest/assets/49083865/386815d5-5b12-44c5-a0a4-9c0ae422fc75)

navneeeth commented:
> Works great, nice job! ![Screenshot 2024-02-08 222209](https://private-user-images.githubusercontent.com/49083865/303551594-386815d5-5b12-44c5-a0a4-9c0ae422fc75.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc1MTQzNDYsIm5iZiI6MTcwNzUxNDA0NiwicGF0aCI6Ii80OTA4Mzg2NS8zMDM1NTE1OTQtMzg2ODE1ZDUtNWIxMi00NGM1LWEwYTQtOWMwYWU0MjJmYzc1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA5VDIxMjcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3YWZkMGQ0YTNhN2Q1NDhmMDU0ZDhjNjY5NWQzZDQwNzVkZjZjZWEzYzMzM2UxZDM0YTQ2Y2Q3ZjBiYmE3ZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KRu2QwE2rA85GKpjO00-QG_OwPAMFW87ubD1B3tpjZw)

Thank you! Appreciate your review!

psharma1984 reviewed (approved):
Tested and its giving the desired results .. 
![Screenshot from 2024-02-18 14-27-40](https://github.com/OneCommunityGlobal/HGNRest/assets/123996264/b2f6af22-365d-41ff-95c7-46b9006f421d)

FyisFe commented on src/helpers/userHelper.js:
```diff
@@ -170,6 +171,14 @@ const userHelper = function () {
           ? `<a href="${mediaUrl}">${mediaUrl}</a>`
           : 'Not provided!';
 
+        const googleDocLinkValue = adminLinks.length > 0
```
Since `adminLinks` is an array, I would add a check to ensure adminLinks is not undefined or null before attempting to access its length property or performing a find operation. This can prevent potential runtime errors if adminLinks is ever undefined.

```js
adminLinks?.length > 0 
```

navneeeth replied:
Hello, @FyisFe! I agree with your suggestion. Will make the update. Thank you!

FyisFe commented on src/helpers/userHelper.js:
```diff
@@ -253,8 +269,8 @@ const userHelper = function () {
       `;
 
       emailSender(
-        'onecommunityglobal@gmail.com, sangam.pravah@gmail.com, onecommunityhospitality@gmail.com',
-        'Weekly Summaries for all active users...',
+        "onecommunityglobal@gmail.com, onecommunityhospitality@gmail.com",
+        "Weekly Summaries for all active users...",
```
Neat picking: I recommend reverting to the original single quotes to maintain consistency throughout our codebase. Could we please keep the quotation style uniform unless there's a specific reason for this change?

navneeeth replied:
No specific reason--will revert to single quotes for maintaining consistency. Good catch! Thank you.

FyisFe commented:
LGTM! It works great.

FyisFe reviewed (approved):
LGTM! It works great.

DiegoSalas27 commented:
Hi Navneeth. I tested your PR and it's working as expected. However, I think we may benefit from logging the weeklySummaries email sending in our local computer (just to notice the email was sent) I added the following video explaining why and how this would be a good addition to the source code (this is out of the scope of your PR, but since you are using a logger, this is a little enhancement for development mode)


https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/470cab99-f556-482a-9894-7fbce4db9420

Great work!

navneeeth commented:
> LGTM! It works great.

Thank you! Appreciate your review.

navneeeth commented:
> Tested and its giving the desired results .. ![Screenshot from 2024-02-18 14-27-40](https://private-user-images.githubusercontent.com/123996264/305757926-b2f6af22-365d-41ff-95c7-46b9006f421d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg3MDcxMTAsIm5iZiI6MTcwODcwNjgxMCwicGF0aCI6Ii8xMjM5OTYyNjQvMzA1NzU3OTI2LWIyZjZhZjIyLTM2NWQtNDFmZi05NWM3LTQ2YjkwMDZmNDIxZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyM1QxNjQ2NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MGI3M2U5NGUzOGJiYTYxZDAyNWNlNGJhYjBjYTMxYTQyOWUxNTMzZmVkOTBlMjdhZDQ1ZWUyYThjZTE1NTEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.jPEArJx3hA6K0PrnkbZPzAVh35ZJftMifGVurrqO5VU)

Thank you! Appreciate your review.

PratimaSingh02 reviewed (changes requested):
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/48998232/130cb912-a542-4d57-8b46-38d943df60e1)
Hi Navneeth, can we have a simple log message like so.
As it might happen that the end user couldn't see the mail but the mail was successfully sent.

Thank You.

Sarthak32 commented:
I have tried testing it but I guess google changed its auth process it shows me :
![Screenshot 2024-02-25 122052](https://github.com/OneCommunityGlobal/HGNRest/assets/92269596/b58bd482-ccdb-4e04-9d09-aa66b0bb0699)

Although code looks good seems to be working for others.

skaush21 reviewed (approved):
Tested weekly summaries email, works great!

* made changes based on the "How to Test" guide
* commented out and uncommented lines of code
* changed sundays value to 3
* added email
* npm install, npn run build, npm run dev
* checked email, received email 
* immediately stopped execution
checked code, looks good!


https://github.com/OneCommunityGlobal/HGNRest/assets/141872402/f399a6a2-129b-44b8-a095-927f487702a0

navneeeth commented:
> Hi Navneeth. I tested your PR and it's working as expected. However, I think we may benefit from logging the weeklySummaries email sending in our local computer (just to notice the email was sent) I added the following video explaining why and how this would be a good addition to the source code (this is out of the scope of your PR, but since you are using a logger, this is a little enhancement for development mode)
> 
>  884f552c-39b0-4296-954e-2911c62fbf02.mp4 
> Great work!

Hello, @DiegoSalas27! Thank you for your suggestion. I do agree that we will benefit from logging the weeklySummaries email sending in our local computer. I have added the recommended enhancement to this PR itself. Many thanks!

navneeeth commented:
> ![image](https://private-user-images.githubusercontent.com/48998232/307516663-130cb912-a542-4d57-8b46-38d943df60e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzODcwNjIsIm5iZiI6MTcxMDM4Njc2MiwicGF0aCI6Ii80ODk5ODIzMi8zMDc1MTY2NjMtMTMwY2I5MTItYTU0Mi00ZDU3LThiNDYtMzhkOTQzZGY2MGUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE0VDAzMjYwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZkYjgyYzZlYjk5OTRjZWFlYjAzM2EzZWIzNzhkMGJhMWIxYmQwNjg2MzIyYjI2MzJiNTZiZmQzYzc5ZWM5ZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZnWpbbEnb_QpHVEKhsFftSo6V2K-TXL3gpFJvUXUNdk) Hi Navneeth, can we have a simple log message like so. As it might happen that the end user couldn't see the mail but the mail was successfully sent.
> 
> Thank You.

Hi, @PratimaSingh02! Thank you for your suggestion. I request you to review the implementation now that I have pushed the necessary updates. Thank you!

navneeeth commented:
> Tested weekly summaries email, works great!
> 
> * made changes based on the "How to Test" guide
> * commented out and uncommented lines of code
> * changed sundays value to 3
> * added email
> * npm install, npn run build, npm run dev
> * checked email, received email
> * immediately stopped execution
>   checked code, looks good!
> 
>  PR.729.mov

Thank you! Appreciate your review.

navneeeth commented:
> I have tried testing it but I guess google changed its auth process it shows me : ![Screenshot 2024-02-25 122052](https://private-user-images.githubusercontent.com/92269596/307572346-b58bd482-ccdb-4e04-9d09-aa66b0bb0699.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzODczMDYsIm5iZiI6MTcxMDM4NzAwNiwicGF0aCI6Ii85MjI2OTU5Ni8zMDc1NzIzNDYtYjU4YmQ0ODItY2NkYi00ZTA0LTlkMDktYWE2NmIwYmIwNjk5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE0VDAzMzAwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFmYzU3OTI5ODlmMTg4YTQ2NmUyOWE5Njk4M2ZhM2JiMjg5ZTI0NjM4YjUwOTI5M2ZkYTVjMjk0N2VlNTQyZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.B6G77RddA7MGukNgW0QLWSsAE_blKJrgWoFC52HYu0c)
> 
> Although code looks good seems to be working for others.

Hello, @Sarthak32! The above error is happening because the email that you are trying to authenticate with is likely not a test user in your Google Cloud Project. Go to your Google Cloud Console and open the project you created for this testing. Navigate to the OAuth consent screen and under a header called 'Test users', click on 'Add Users' and add the above email. You should then be able to set up the emailing feature. It will help you review PRs related to email sending/receiving and also work on features related to it. Hope this helps. Thank you!

DiegoSalas27 commented:
> > Hi Navneeth. I tested your PR and it's working as expected. However, I think we may benefit from logging the weeklySummaries email sending in our local computer (just to notice the email was sent) I added the following video explaining why and how this would be a good addition to the source code (this is out of the scope of your PR, but since you are using a logger, this is a little enhancement for development mode)
> 
> > 
> 
> >  884f552c-39b0-4296-954e-2911c62fbf02.mp4 
> 
> > Great work!
> 
> 
> 
> Hello, @DiegoSalas27! Thank you for your suggestion. I do agree that we will benefit from logging the weeklySummaries email sending in our local computer. I have added the recommended enhancement to this PR itself. Many thanks! 

Awesome. You are welcome.

one-community commented:
Thank you all, moving this to final review!

cvtqx reviewed (approved):
Everything works as described, the email sent includes the google doc link. And by the way, excellent instructions and video for the PR! 

![Screen Shot 2024-03-22 at 11 31 35 AM](https://github.com/OneCommunityGlobal/HGNRest/assets/107726913/de216eba-8546-4b82-b5f6-afb02888c8ef)

navneeeth commented:
> Everything works as described, the email sent includes the google doc link. And by the way, excellent instructions and video for the PR!
> 
> ![Screen Shot 2024-03-22 at 11 31 35 AM](https://private-user-images.githubusercontent.com/107726913/316080454-de216eba-8546-4b82-b5f6-afb02888c8ef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMjIwODAsIm5iZiI6MTcxMTEyMTc4MCwicGF0aCI6Ii8xMDc3MjY5MTMvMzE2MDgwNDU0LWRlMjE2ZWJhLTg1NDYtNGI4Mi1iNWY2LWFmYjAyODg4YzhlZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyMlQxNTM2MjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kN2IwZGEzZmUwZmIyOTVhNTJjM2Q5ZWU4YTlkNTEwODg0ZTAwZmM2ODRiNjA2NzNlYWM5NDI5ODgxODhmNmYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.f_Gr4TprSiq1Y3z8BDMGCGgUoUeeezmjK3kAhOvQV0o)

Thank you for your review, @cvtqx! Appreciate it!

one-community commented:
Thank you all, merging!
